rc/src/lib.rs

```diff
@@ -0,0 +1,135 @@
+use std::ptr;
+use std::mem;
+use std::clone::Clone;
+use std::ops::{Deref,Drop};
+
+// the heap object
+struct RcCore<T> {
+    count: usize,
+    wcount: usize, // number of weakrefs
+    val: Option<T>
```

Comment:
So this layout definitely works, but a bit of extra space it used by the discriminant of the option. You could do some clever things to avoid using the option.